src/main/integrations/discord-presence/index.ts

```diff
@@ -78,93 +64,78 @@ function stringLimit(str: string, limit: number, minimum: number) {
-        this.lastDuration = duration;
-        this.lastTimeSeconds = timeSeconds;
-      }
+      const oldState = this.videoState ?? null;
+      const oldProgress = this.progress ?? null;
+      const oldId = this.videoDetails?.id ?? null;
+
+      this.videoDetails = {
+        title: state.videoDetails.title,
+        author: state.videoDetails.author,
+        album: state.videoDetails.album,
+        id: state.videoDetails.id,
+        thumbnails: state.videoDetails.thumbnails,
+        durationSeconds: state.videoDetails.durationSeconds
+      };
+      this.progress = Math.floor(state.videoProgress);
+      this.videoState = state.trackState;
 
-      this.endTimestamp = state.trackState === VideoState.Playing ? adjustedTimeSeconds + (state.videoDetails.durationSeconds - videoProgress) : undefined;
       if (
-        this.lastEndTimestamp !== this.endTimestamp ||
-        state.videoDetails.title !== this.lastVideoDetailsTitle ||
-        state.videoDetails.author !== this.lastVideoDetailsAuthor ||
-        state.videoDetails.id !== this.lastVideoDetailsId ||
-        state.trackState !== this.lastTrackState
-      ) {
-        this.lastEndTimestamp = this.endTimestamp;
-        this.lastVideoDetailsTitle = state.videoDetails.title;
-        this.lastVideoDetailsAuthor = state.videoDetails.author;
-        this.lastVideoDetailsId = state.videoDetails.id;
-        this.lastTrackState = state.trackState;
-
-        const thumbnail = getHighestResThumbnail(state.videoDetails.thumbnails);
-        this.discordClient.setActivity({
-          // Discord accepts Playing, Listening, and Watching. But ignores it and sets it to 0 (Playing)
-          // We'll still send a type in case Discord at some point updates to allow this
-          type: DiscordActivityType.Listening,
-          details: stringLimit(state.videoDetails.title, 128, 2),
-          state: stringLimit(state.videoDetails.author, 128, 2),
-          timestamps: {
-            end: state.trackState === VideoState.Playing ? this.endTimestamp : undefined
-          },
-          assets: {
-            large_image: thumbnail && thumbnail.length <= 256 ? thumbnail : "ytmd-logo",
-            large_text: stringLimit(state.videoDetails.title, 128, 2),
-            small_image: getSmallImageKey(state.trackState),
-            small_text: getSmallImageText(state.trackState)
-          },
-          instance: false,
-          buttons: [
-            {
-              label: "Play on YouTube Music",
-              url: `https://music.youtube.com/watch?v=${state.videoDetails.id}`
-            },
-            {
-              label: "Play on YouTube Music Desktop",
-              url: `ytmd://play/${state.videoDetails.id}`
-            }
-          ]
-        });
-      }
+        this.videoDetails &&
+        this.videoDetails.id &&
+        (oldState !== this.videoState || Math.abs(oldProgress - this.progress) > 5 || oldId !== this.videoDetails.id)
+      )
+        this.setActivity();
```

Comment:
I feel like this might need to be re-done. There's 2 primary things to address here including.

I feel like this doesn't read very well at a initial glance of the code. I struggled to understand with the variable names all closely matching up from the `state.videoDetails` and `this.videoDetails` having the property names (but they do store the same thing)


Secondly, line 136 doing the comparisons, does this only check for a change of 5 seconds? Does me constantly rewinding it 1 second mean that my Discord RPC is going to desync now?
While the old method was a bit more complex I feel it is needed.


This might be something which I play around with directly in your Branch to work out.

Reply:
From my own research the state changes whenever you move the progress manually. I have yet to find an instance where checking the id and the video state didnt work. But I believe it could be different for people with different setups

Reply:
Do you have a Premium account?

Because one example to try would be downloaded music.